stellargraph/mapper/link_mappers.py

```diff
@@ -163,6 +162,120 @@ def on_epoch_end(self):
+                This class is responsible for calling the right type of walk generator and then return batch_size of sample source and target pairs  from those walks.
+                These (target, context) pairs are to be used to train or inference, and the embeddings
+                calculated for the links created by a binary operator applied to the target and context nodes,
+                are passed to the downstream task of link prediction or link attribute inference.
+                The target and context nodes of the links are used as head nodes for which subgraphs are sampled.
+                The subgraphs are sampled from all nodes.
+
+  """
+
+    def __init__(self, generator, walker):
+
+        self.generator = generator  # graphlinkgenerator instance
+
+        self.ids = (
+            []
+        )  # Since this is an instance of on demand sampling, at the initialization we don't have the pregenerated head samples.
+
+        if isinstance(walker, UnsupervisedSampler):
+
+            self.walker = walker
+
+            self.data_size = (
+                len(self.walker.nodes)
+                * self.walker.length
+                * self.walker.number_of_walks
+            )  # an estimate of the  upper bound on how many samples are generated in each epoch
+
+            print(
+                "Running GraphSAGELinkGenerator with an estimated {} batches generated on the fly per epoch.".format(
+                    round(self.data_size / self.generator.batch_size)
+                )
+            )
+
+            self._gen = self.walker.generator(
+                self.generator.batch_size
+            )  # the generator method from the sampler with the batch-size from the link generator method
+        else:
+            raise TypeError(
+                "({}) UnsupervisedSampler is required.".format(type(self).__name__)
+            )
+
+    def __getitem__(self, batch_num):
+        """
+        Generate one batch of data.
+
+        Args:
+            batch_num<int>: number of a batch
+
+        Returns:
+            batch_feats<list>: Node features for nodes and neighbours sampled from a
+                batch of the supplied IDs
+            batch_targets<list>: Targets/labels for the batch.
+
+        """
+
+        if batch_num >= self.__len__():
+            raise IndexError(
+                "Mapper: batch_num larger than number of esstaimted  batches for this epoch."
+            )
+        # print("Fetching {} batch {} [{}]".format(self.name, batch_num, start_idx))
+
+        # Get head nodes and labels
+        head_ids, batch_targets = next(self._gen)
+        self.ids = list(head_ids)
+
+        # Get head node types from all src, dst nodes extracted from all links,
+        # and make sure there's only one pair of node types:
+        self.head_node_types = self._infer_head_node_types(self.generator.schema)
```

Comment:
Inferring the head node types & sampling schema every time a batch is generated is not the best. Is there a way we can do this just once?

Reply:
I agree, this should be done at construction time (in __init__)

Reply:
Now, the head_node_types and sampling schema are inferred only once at the time when the first batch is sampled.